controllers/cps_controller_helper.go

```diff
@@ -0,0 +1,303 @@
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/utils/strings/slices"
+	ctrl "sigs.k8s.io/controller-runtime"
+	cacheRuntime "sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/healthz"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/metrics/server"
+)
+
+var (
+	setupLog       = ctrl.Log.WithName("setup")
+	errCrdNotReady = errors.New("CRD is not ready")
+)
+
+func (r *GatekeeperReconciler) handleCPSController(mainCtx context.Context,
+	gatekeeper *operatorv1alpha1.Gatekeeper,
+) error {
+	isCRDReady, err := checkCPSCrdAvailable(mainCtx, r.DynamicClient)
+	if err != nil {
+		return err
+	}
+
+	if !isCRDReady {
+		return errCrdNotReady
+	}
+
+	isAutomaticOn, err := checkCPScontrollerPrereqs(gatekeeper, r.DynamicClient)
+	if err != nil {
+		return err
+	}
+
+	// auditFromCache is not set to Automatic, so stop the existing ConstraintPodStatus controller
+	if !isAutomaticOn {
+		if r.cpsCtrlCtx != nil {
+			r.StopCPSController()
+		}
+
+		return nil
+	}
+
+	if r.isCPSCtrlRunning {
+		return nil
+	}
+
+	r.cpsCtrlCtx, r.cpsCtrlCtxCancel = context.WithCancel(mainCtx)
+
+	cpsMgr, err := ctrl.NewManager(r.KubeConfig, ctrl.Options{
+		Scheme: r.Scheme,
+		Metrics: server.Options{
+			BindAddress: ":8083",
+		},
+		HealthProbeBindAddress: ":8084",
+		LeaderElection:         r.EnableLeaderElection,
+		LeaderElectionID:       "5ff985ccc.constraintstatuspod.gatekeeper.sh",
+		Cache: cacheRuntime.Options{
+			ByObject: map[client.Object]cacheRuntime.ByObject{
+				&gkv1beta1.ConstraintPodStatus{}: {
+					Transform: func(obj interface{}) (interface{}, error) {
+						constraintStatus := obj.(*gkv1beta1.ConstraintPodStatus)
+						// Only cache fields that are utilized by the controllers.
+						guttedObj := &gkv1beta1.ConstraintPodStatus{
+							TypeMeta: constraintStatus.TypeMeta,
+							ObjectMeta: metav1.ObjectMeta{
+								Name:      constraintStatus.Name,
+								Labels:    constraintStatus.Labels,
+								Namespace: constraintStatus.Namespace,
+							},
+							Status: gkv1beta1.ConstraintPodStatusStatus{
+								ObservedGeneration: constraintStatus.Status.ObservedGeneration,
+								Operations:         constraintStatus.Status.Operations,
+							},
+						}
+
+						return guttedObj, nil
+					},
+				},
+			},
+		},
+	})
+	if err != nil {
+		setupLog.Error(err, "Failed to setup NewManager for ConstraintPodStatus contoller")
+
+		return err
+	}
+
+	// Use another go routine for the ConstraintPodStatus controller
+	go func() {
+		for {
+			// When controller is running then exit this loop
+			if r.isCPSCtrlRunning {
+				return
+			}
+
+			err = r.addConstraintPodStatusManager(mainCtx, cpsMgr)
+			if err != nil {
+				r.isCPSCtrlRunning = false
```

Comment:
By setting this to false and while this is sleeping for 3 seconds, a Reconcile request could come in for the Gatekeeper resource, and `handleCPSController` could be called again because it sees that it's running and it isn't running.

Reply:
Is it better to add `sync.Mutex` around `r.isCPSCtrlRunning`?

Reply:
I'll send a diff in a little bit with I think a simpler approach.

Reply:
Maybe it's not simpler after all lol. Anyways, it makes it so that if the CPS manager exits due to an error, it triggers a reconcile of the Gatekeeper resource. I think I've also managed to guard against race conditions as well by setting `r.isCPSCtrlRunning` to `true` and `r.isCPSCtrlRunning` to `false` in only one place.

```diff
diff --git a/controllers/cps_controller_helper.go b/controllers/cps_controller_helper.go
index 9f1cabd..af91937 100644
--- a/controllers/cps_controller_helper.go
+++ b/controllers/cps_controller_helper.go
@@ -5,6 +5,7 @@ import (
        "strings"
        "time"
 
+       "github.com/gatekeeper/gatekeeper-operator/api/v1alpha1"
        operatorv1alpha1 "github.com/gatekeeper/gatekeeper-operator/api/v1alpha1"
        gkv1alpha1 "github.com/open-policy-agent/gatekeeper/v3/apis/config/v1alpha1"
        "github.com/open-policy-agent/gatekeeper/v3/apis/status/v1beta1"
@@ -19,8 +20,7 @@ import (
        ctrl "sigs.k8s.io/controller-runtime"
        cacheRuntime "sigs.k8s.io/controller-runtime/pkg/cache"
        "sigs.k8s.io/controller-runtime/pkg/client"
-       "sigs.k8s.io/controller-runtime/pkg/healthz"
-       "sigs.k8s.io/controller-runtime/pkg/manager"
+       "sigs.k8s.io/controller-runtime/pkg/event"
        "sigs.k8s.io/controller-runtime/pkg/metrics/server"
 )
 
@@ -48,7 +48,7 @@ func (r *GatekeeperReconciler) handleCPSController(mainCtx context.Context,
 
        // auditFromCache is not set to Automatic, so stop the existing ConstraintPodStatus controller
        if !isAutomaticOn {
-               if r.cpsCtrlCtx != nil {
+               if r.isCPSCtrlRunning {
                        r.StopCPSController()
                }
 
@@ -59,7 +59,9 @@ func (r *GatekeeperReconciler) handleCPSController(mainCtx context.Context,
                return nil
        }
 
-       r.cpsCtrlCtx, r.cpsCtrlCtxCancel = context.WithCancel(mainCtx)
+       var cpsCtrlCtx context.Context
+
+       cpsCtrlCtx, r.cpsCtrlCtxCancel = context.WithCancel(mainCtx)
 
        cpsMgr, err := ctrl.NewManager(r.KubeConfig, ctrl.Options{
                Scheme: r.Scheme,
@@ -100,29 +102,45 @@ func (r *GatekeeperReconciler) handleCPSController(mainCtx context.Context,
                return err
        }
 
-       // Use another go routine for the ConstraintPodStatus controller
-       go func() {
-               for {
-                       // When controller is running then exit this loop
-                       if r.isCPSCtrlRunning {
-                               return
-                       }
+       constraintToSyncOnly := r.getConstraintToSyncOnly(mainCtx)
 
-                       err = r.addConstraintPodStatusManager(mainCtx, cpsMgr)
-                       if err != nil {
-                               r.isCPSCtrlRunning = false
+       if err := (&ConstraintPodStatusReconciler{
+               Scheme:               r.Scheme,
+               Client:               cpsMgr.GetClient(),
+               DynamicClient:        r.DynamicClient,
+               Log:                  ctrl.Log.WithName("ConstraintPodStatus"),
+               Namespace:            r.Namespace,
+               ConstraintToSyncOnly: constraintToSyncOnly,
+               DiscoveryStorage:     r.DiscoveryStorage,
+       }).SetupWithManager(cpsMgr); err != nil {
+               setupLog.Error(err, "unable to create controller", "controller", "ConstraintPodStatus")
 
-                               setupLog.Error(err, "unable to start ConstraintPodStatus manager, try again after 3 second")
-                               time.Sleep(3 * time.Second)
+               return err
+       }
 
-                               continue
-                       }
+       r.isCPSCtrlRunning = true
 
-                       // Child context closed
-                       <-r.cpsCtrlCtx.Done()
-                       r.isCPSCtrlRunning = false
+       // Use another go routine for the ConstraintPodStatus controller
+       go func() {
+               err := cpsMgr.Start(cpsCtrlCtx)
+               defer r.cpsCtrlCtxCancel()
+               r.cpsCtrlCtxCancel = nil
+               r.isCPSCtrlRunning = false
 
-                       return
+               if err != nil {
+                       setupLog.Error(err, "A problem running ConstraintPodStatus manager. Triggering a reconcile to restart it.")
+
+                       r.ManualReconcileTrigger <- event.GenericEvent{
+                               Object: &unstructured.Unstructured{
+                                       Object: map[string]interface{}{
+                                               "apiVersion": v1alpha1.GroupVersion.String(),
+                                               "kind":       "Gatekeeper",
+                                               "metadata": map[string]interface{}{
+                                                       "name": defaultGatekeeperCrName,
+                                               },
+                                       },
+                               },
+                       }
                }
        }()
 
@@ -186,47 +204,6 @@ func (r *GatekeeperReconciler) getConstraintToSyncOnly(mainCtx context.Context)
        return constraintToSyncOnly
 }
 
-func (r *GatekeeperReconciler) addConstraintPodStatusManager(mainCtx context.Context, mgr manager.Manager,
-) error {
-       constraintToSyncOnly := r.getConstraintToSyncOnly(mainCtx)
-
-       if err := (&ConstraintPodStatusReconciler{
-               Scheme:               r.Scheme,
-               Client:               mgr.GetClient(),
-               DynamicClient:        r.DynamicClient,
-               Log:                  ctrl.Log.WithName("ConstraintPodStatus"),
-               Namespace:            r.Namespace,
-               ConstraintToSyncOnly: constraintToSyncOnly,
-               DiscoveryStorage:     r.DiscoveryStorage,
-       }).SetupWithManager(mgr); err != nil {
-               setupLog.Error(err, "unable to create controller", "controller", "ConstraintPodStatus")
-
-               return err
-       }
-
-       if err := mgr.AddHealthzCheck("healthz", healthz.Ping); err != nil {
-               setupLog.Error(err, "unable to set up health check")
-
-               return err
-       }
-
-       if err := mgr.AddReadyzCheck("readyz", healthz.Ping); err != nil {
-               setupLog.Error(err, "unable to set up ready check")
-
-               return err
-       }
-
-       r.isCPSCtrlRunning = true
-
-       if err := mgr.Start(r.cpsCtrlCtx); err != nil {
-               setupLog.Error(err, "problem running ConstraintPodStatus manager")
-
-               return err
-       }
-
-       return nil
-}
-
 // Check ConstraintPodStatus Crd status is "True" and type is "NamesAccepted"
 func checkCPSCrdAvailable(mainCtx context.Context, dynamicClient *dynamic.DynamicClient) (bool, error) {
        crdGVR := schema.GroupVersionResource{
@@ -296,8 +273,17 @@ func checkCPScontrollerPrereqs(gatekeeper *operatorv1alpha1.Gatekeeper, dynamicC
 }
 
 func (r *GatekeeperReconciler) StopCPSController() {
+       if r.cpsCtrlCtxCancel == nil {
+               return
+       }
+
+       setupLog.Info("Gatekeeper auditFromCache unset from Automatic. Stopping the ConstraintPodStatus manager.")
+
        r.cpsCtrlCtxCancel()
-       r.isCPSCtrlRunning = false
 
-       setupLog.Info("Gatekeeper deleted. Stopping the ConstraintPodStatus manager.")
+       for r.isCPSCtrlRunning {
+               setupLog.Info("Waiting for the ConstraintPodStatus manager to shutdown")
+
+               time.Sleep(1 * time.Second)
+       }
 }
diff --git a/controllers/gatekeeper_controller.go b/controllers/gatekeeper_controller.go
index f3f6ae3..e503e6b 100644
--- a/controllers/gatekeeper_controller.go
+++ b/controllers/gatekeeper_controller.go
@@ -39,7 +39,9 @@ import (
        "sigs.k8s.io/controller-runtime/pkg/client"
        "sigs.k8s.io/controller-runtime/pkg/controller"
        "sigs.k8s.io/controller-runtime/pkg/event"
+       "sigs.k8s.io/controller-runtime/pkg/handler"
        "sigs.k8s.io/controller-runtime/pkg/predicate"
+       "sigs.k8s.io/controller-runtime/pkg/source"
 
        operatorv1alpha1 "github.com/gatekeeper/gatekeeper-operator/api/v1alpha1"
        "github.com/gatekeeper/gatekeeper-operator/controllers/merge"
@@ -131,13 +133,13 @@ type GatekeeperReconciler struct {
        Namespace    string
        PlatformInfo platform.PlatformInfo
        // Flag for ConstraintPodStatus controller installation
-       DiscoveryStorage     *DiscoveryStorage
-       isCPSCtrlRunning     bool
-       DynamicClient        *dynamic.DynamicClient
-       KubeConfig           *rest.Config
-       EnableLeaderElection bool
-       cpsCtrlCtx           context.Context //nolint:containedctx
-       cpsCtrlCtxCancel     context.CancelFunc
+       DiscoveryStorage       *DiscoveryStorage
+       ManualReconcileTrigger chan event.GenericEvent
+       isCPSCtrlRunning       bool
+       DynamicClient          *dynamic.DynamicClient
+       KubeConfig             *rest.Config
+       EnableLeaderElection   bool
+       cpsCtrlCtxCancel       context.CancelFunc
 }
 
 type crudOperation uint32
@@ -204,7 +206,7 @@ func (r *GatekeeperReconciler) Reconcile(ctx context.Context, req ctrl.Request)
        err := r.Get(ctx, req.NamespacedName, gatekeeper)
        if err != nil {
                if apierrors.IsNotFound(err) {
-                       if r.cpsCtrlCtx != nil && r.isCPSCtrlRunning {
+                       if r.isCPSCtrlRunning {
                                r.StopCPSController()
                        }
 
@@ -239,7 +241,7 @@ func (r *GatekeeperReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 }
 
 // SetupWithManager sets up the controller with the Manager.
-func (r *GatekeeperReconciler) SetupWithManager(mgr ctrl.Manager) error {
+func (r *GatekeeperReconciler) SetupWithManager(mgr ctrl.Manager, fromCPSMgrSource *source.Channel) error {
        return ctrl.NewControllerManagedBy(mgr).
                WithOptions(controller.Options{MaxConcurrentReconciles: int(1)}).
                For(&operatorv1alpha1.Gatekeeper{}).
@@ -251,6 +253,7 @@ func (r *GatekeeperReconciler) SetupWithManager(mgr ctrl.Manager) error {
                                return oldGeneration != newGeneration
                        },
                }).
+               WatchesRawSource(fromCPSMgrSource, &handler.EnqueueRequestForObject{}).
                Complete(r)
 }
 
diff --git a/main.go b/main.go
index 96f3fe7..5199121 100644
--- a/main.go
+++ b/main.go
@@ -27,7 +27,9 @@ import (
        "k8s.io/client-go/kubernetes"
        _ "k8s.io/client-go/plugin/pkg/client/auth"
 
+       "sigs.k8s.io/controller-runtime/pkg/event"
        "sigs.k8s.io/controller-runtime/pkg/metrics/server"
+       "sigs.k8s.io/controller-runtime/pkg/source"
        "sigs.k8s.io/controller-runtime/pkg/webhook"
 
        operatorv1alpha1 "github.com/gatekeeper/gatekeeper-operator/api/v1alpha1"
@@ -123,20 +125,24 @@ func main() {
 
        dynamicClient := dynamic.NewForConfigOrDie(mgr.GetConfig())
 
+       manualReconcileTrigger := make(chan event.GenericEvent, 1024)
+       fromCPSMgrSource := &source.Channel{Source: manualReconcileTrigger, DestBufferSize: 1024}
+
        if err = (&controllers.GatekeeperReconciler{
-               Client:               mgr.GetClient(),
-               Log:                  ctrl.Log.WithName("controllers").WithName("Gatekeeper"),
-               Scheme:               mgr.GetScheme(),
-               Namespace:            namespace,
-               PlatformInfo:         platformInfo,
-               DynamicClient:        dynamicClient,
-               KubeConfig:           cfg,
-               EnableLeaderElection: enableLeaderElection,
+               Client:                 mgr.GetClient(),
+               Log:                    ctrl.Log.WithName("controllers").WithName("Gatekeeper"),
+               Scheme:                 mgr.GetScheme(),
+               Namespace:              namespace,
+               PlatformInfo:           platformInfo,
+               DynamicClient:          dynamicClient,
+               KubeConfig:             cfg,
+               EnableLeaderElection:   enableLeaderElection,
+               ManualReconcileTrigger: manualReconcileTrigger,
                DiscoveryStorage: &controllers.DiscoveryStorage{
                        Log:       ctrl.Log.WithName("discovery_storage"),
                        ClientSet: kubernetes.NewForConfigOrDie(mgr.GetConfig()),
                },
-       }).SetupWithManager(mgr); err != nil {
+       }).SetupWithManager(mgr, fromCPSMgrSource); err != nil {
                setupLog.Error(err, "unable to create controller", "controller", "Gatekeeper")
                os.Exit(1)
        }
```

Reply:
oh my god Thanks

Reply:
@mprahl  one thing after `setupWithManager` starts successfully, it won't reach ` r.isCPSCtrlRunning = true`

Reply:
@yiraeChristineKim no problem! I think on second thought, it's probably safer to always trigger a reconcile after the CPS manager/controller exits, regardless of if there was an error. It might be slightly wasteful but it shouldn't happen often and it will take away all chances of encountering a race condition.

Reply:
@yiraeChristineKim , yeah, the defer isn't strictly necessary. I thought setting to nil would be helpful though to return it to the original state before it ran for the first time.

This isn't required though as long as nothing is checking `cpsCtrlCtxCancel` to see if it's defined to determine if the controller is running, which it shouldn't be anymore after the latest changes.